### PR TITLE
CSHARP-1004 Add client side column encryption

### PR DESCRIPTION
**Pull request by joao-r-reis**

### CODE REVIEW

**Comment by joao-r-reis** [comment]
@absurdfarce I'm still working on the api docs and driver docs justfyi but the actual code is ready to be reviewed. All tests are passing.

File: src/Cassandra/Policies/IColumnEncryptionPolicy.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+        /// <summary>
+        /// Decrypt the specified (encrypted) bytes using the cryptography materials.
+        /// This method is used by the driver internally before providing the results to the application.
+        /// </summary>
+        /// <returns>Decrypted data in a byte array.</returns>
+        byte[] Decrypt(object key, byte[] encryptedBytes);
+
+        /// <summary>
+        /// Retrieves the cryptography materials for the specified column. If the column is not encrypted then should return 'null'.
```
Comment (emeliawilkinson24) [inline src/Cassandra/Policies/IColumnEncryptionPolicy.cs:49]:
```suggestion
        /// Retrieves the cryptography materials for the specified column. If the column is not encrypted then it should return 'null'.
```

**Review by emeliawilkinson24 — APPROVED** [review]
Some optional grammatical changes.

File: doc/features/column-encryption/README.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
+# Column Encryption
+
+## Overview
+
+Support for client-side encryption of data was added in version 3.20.0 of the C# driver. When using this feature data will be encrypted on-the-fly according to a provided implementation of `IColumnEncryptionPolicy`. This policy is also used to decrypt data in returned rows.  If a prepared statement is used this decryption is transparent to the user; retrieved data will be decrypted and converted into the original
+type (according to definitions in the encryption policy). Support for simple (i.e. non-prepared) statements is also available, although in this case values must be wrapped using the `EncryptedValue` type.
+
+Client-side encryption and decryption should work against all versions of Cassandra, DSE and Astra. It does not utilize any server-side functionality to do its work.
```
[the remark this answers is not in the document]
Comment (joao-r-reis) [reply doc/features/column-encryption/README.md:8]: Are you sure this is better? Doesn't look right to me for some reason

File: doc/features/column-encryption/README.md [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+var cluster =
+    Cluster.Builder()
+        .AddContactPoint("127.0.0.1")
+        .WithColumnEncryptionPolicy(policy)
+        .Build();
+```
+
+### AESColumnEncryptionPolicy
+
+`AESColumnEncryptionPolicy` is an implementation of `IColumnEncryptionPolicy` which provides encryption and decryption via AES-128, AES-192 or AES-256 according to the size of the key that is provided. This class is currently the only available column encryption policy implementation, although users can certainly implement their own by implementing `IColumnEncryptionPolicy` or subclassing the abstract `BaseColumnEncryptionPolicy` class which provides some out of the box functionality to manage the encrypted column metadata.
```
[the remark this answers is not in the document]
Comment (joao-r-reis) [reply doc/features/column-encryption/README.md:25]: Same as above

File: doc/features/column-encryption/README.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+To mark a column as "encrypted" users need to call the `AddColumn` method. This method has two overloads, one of them has an additional `IColumnInfo` parameter which should only be used if the column has a type of `list`, `set`, `map`, `udt`, `tuple` or `custom`. If you are using one of these six types on that particular column then you need to provide a `IColumnInfo` object with a type that matches the cql type (e.g. `list` requires a `ListColumnInfo` object to be provided in the `AddColumn` method).
```
[the remark this answers is not in the document]
Comment (joao-r-reis) [reply doc/features/column-encryption/README.md:27]: The other one is described in the next paragraph

**Pull request merged**